RIOT-OS/RIOT
# esp8266: esp_wifi netdev driver

### Contribution description

This PR adds a `netdev` driver for infrastructure mode WiFi for ESP8266 using the built-in WiFi module. Using this `netdev` driver, ESP8266 can use an existing WiFi network to get connectivity. The `netdev` driver supports WPA authentication only.

The Wifi network interface (module `esp_wifi`) and the ESP-NOW network interface (module `esp_now` #9917)
can be used simultaneously, for example, to realize a border router for a mesh network which uses ESP-NOW.

### Testing procedure

- Compile and flash `example/gnrc_networking` to at least one ESP8266 node using your AP configuration, e.g.,
    ```
    CFLAGS='-DESP_WIFI_SSID=\"<your SSID>\" -DESP_WIFI_PASS=\"your passphrase\"' USEMODULE="esp_wifi" make BOARD=esp8266-esp-12x -C examples/gnrc_networking flash
    ```
- Once the application is flashed connect the ESP32 node with any terminal programm with a default baudrate of 115.200 bps.
- Use `ifconfig` command to check the existence of the network interface. You should be able to observe an output as follows:
   ```
   Iface  7  HWaddr: 5C:CF:7F:80:3F:08  Link: up 
          MTU:1500  HL:64  RTR  
          RTR_ADV  
          Source address length: 6
          Link type: wireless
          inet6 addr: fe80::5ecf:7fff:fe80:3f08  scope: local  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ff80:3f08
          inet6 group: ff02::1a
          
          Statistics for Layer 2
            RX packets 12  bytes 2698
            TX packets 1 (Multicast: 1)  bytes 0
            TX succeeded 0 errors 0
          Statistics for IPv6
            RX packets 0  bytes 0
            TX packets 1 (Multicast: 1)  bytes 64
            TX succeeded 1 errors 0
    ```
- Ping any host in LAN.
- If the router provides a global routing prefix, check with `ifconfig` that the prefix is set.
   ```
   Iface  7  HWaddr: 5C:CF:7F:80:3F:08  Link: up 
          MTU:1440  HL:255  RTR  
          RTR_ADV  
          Source address length: 6
          Link type: wireless
          inet6 addr: fe80::5ecf:7fff:fe80:3f08  scope: local  VAL
          inet6 addr: 2003:c1:e72f:6c40:5ecf:7fff:fe80:3f08  scope: global  VAL
          ...
   ```
- Stress testing from LAN with `sudo ping6 fe80::<ESP_IID> -Ieth0 -s1392 -i 0.0001` should be stable over hours.
- It should also be possible to ping simultaneously
   - the ESP8266 node from the LAN with `sudo ping6 fe80::<ESP_IID> -Ieth0 -s1392 -i 0.0001`
   - the node in the LAN swith `ping6 1000 fe80::<LAN_IID> 1392 0` .

### Issues/PRs references

This PR depends on PRs #10656 and #10766 and is related to PR #9917.
